pysph/tools/ipy_viewer.py

```diff
@@ -572,7 +581,684 @@ def _stride_and_component(self, data, pa_widgets):
+            pa_widgets = self._widgets.particles[array_name]
+            if (pa_widgets.scalar.value != 'None' and
+                    pa_widgets.is_visible.value is True):
+
+                n = pa_widgets.masking_factor.value
+                stride, component = self._stride_and_component(
+                    temp_data[array_name], pa_widgets
+                )
+                c = self._get_c(
+                    pa_widgets,
+                    temp_data[array_name],
+                    component,
+                    stride
+                )
+                if self.xmin is None:
+                    self.xmin = min(temp_data[array_name].x)
+                elif min(temp_data[array_name].x) < self.xmin:
+                    self.xmin = min(temp_data[array_name].x)
+                if self.xmax is None:
+                    self.xmax = max(temp_data[array_name].x)
+                elif max(temp_data[array_name].x) > self.xmax:
+                    self.xmax = max(temp_data[array_name].x)
+                min_c = min(c)
+                max_c = max(c)
+                llim = pa_widgets.right_spine_lower_lim.value
+                ulim = pa_widgets.right_spine_upper_lim.value
+                if llim != '':
+                    min_c = float(llim)
+                if ulim != '':
+                    max_c = float(ulim)
+                if max_c-min_c != 0:
+                    c_norm = (c - min_c)/(max_c - min_c)
+                elif max_c != 0:
+                    c_norm = c/max_c
+                else:
+                    c_norm = c
+
+                color = 'C' + str(
+                    list(
+                        self._widgets.particles.keys()
+                    ).index(array_name)
+                )
+                ax = self._scatters_ax['host'].twinx()
+                self._scatters[array_name] = ax.scatter(
+                    temp_data[array_name].x[component::stride][::n],
+                    c_norm[::n],
+                    color=color,
+                    label=array_name,
+                )
+                ax.set_ylim(-0.1, 1.1)
+                self._make_patch_spines_invisible(ax)
+                self._scatters_ax[array_name] = ax
+
+        self._scatters_ax['host'].set_xlim(self.xmin-0.1, self.xmax+0.1)
+        self._make_patch_spines_invisible(self._scatters_ax['host'])
+
+        self.solver_time_textbox = None
+        # So that _show_solver_time_handler does not glitch at intialization.
+        self._frame_handler(None)
+        self._show_solver_time_handler(None)
+        self._right_spine_handler(None)
+
+    def _make_patch_spines_invisible(self, ax, val=True):
+        '''
+        Helper function for making individual y-axes
+        for different particle arrays
+        '''
+        ax.set_frame_on(val)
+        ax.patch.set_visible(False)
+        for sp in ax.spines.values():
+            sp.set_visible(False)
+
+    def _frame_handler(self, change):
```

Comment:
There seems to be a lot of repeated code here and in the other Viewers.  Is there a way these can be abstracted out to minimize code repetition?

Reply:
I've refactored as much code as I could for the 2D and 3D Viewers, but the problem still remains that all three viewers use markedly different ways to display the plots. Not just this, the biggest issue I've found in refactoring code is that the way the legends are handled for the three viewers are completely different, and thus cannot be easily abstracted out.

Reply:
OK.